src/main/java/net/glowstone/block/itemtype/ItemBow.java

```diff
@@ -0,0 +1,101 @@
+package net.glowstone.block.itemtype;
+
+import java.util.concurrent.ThreadLocalRandom;
+import net.glowstone.entity.GlowPlayer;
+import net.glowstone.util.InventoryUtil;
+import org.bukkit.GameMode;
+import org.bukkit.Material;
+import org.bukkit.enchantments.Enchantment;
+import org.bukkit.entity.Arrow;
+import org.bukkit.entity.SpectralArrow;
+import org.bukkit.entity.TippedArrow;
+import org.bukkit.inventory.ItemStack;
+
+public class ItemBow extends ItemTimedUsage {
+    private static final long TICKS_TO_FULLY_CHARGE = 20;
+    private static final double MAX_BASE_DAMAGE = 9;
+    private static final double MAX_SPEED = 53;
+    private ItemStack currentArrow;
+
+    @Override
+    public void startUse(GlowPlayer player, ItemStack item) {
+        findArrow: for (ItemStack itemStack : player.getInventory().getContents()) {
+            switch (itemStack.getType()) {
+                case ARROW:
+                    if (currentArrow == null) {
+                        currentArrow = itemStack;
```

Comment:
I don't think this is how it should be implemented. The arrow lookup should be done in the `endUse`: the startUse should just check if there *is* an arrow somewhere.

Suppose that the player has two stacks of arrows in their inventory. This function will determine that the first stack is to be used, and will be stored in `currentArrow`. If the stack is dropped or removed during the bow pulling (by a plugin or whatever), the `endUse` function won't be able to find the stack with this current system. This search should be done when by the `endUse`.

The `endUse` function should find the arrow and get the slot index for the stack. Then, it will summon the arrow and remove one/delete and update the slot (`inventory.setItem(slot, arrowStack)`)

Reply:
Done in cb38717fadd2b1bdbaa3f0108b47cff6680c9c9d